src/main/scala/com/fulcrumgenomics/bam/ConsensusCaller.scala

```diff
@@ -0,0 +1,369 @@
+) extends Iterator[SAMRecord] {
+  import ConsensusCaller.ReadType
+  import ConsensusCaller.ReadType._
+  import ConsensusCaller._
+
+  private val iter = input.buffered
+  private val nextConsensusRecords: mutable.Queue[SAMRecord] = mutable.Queue[SAMRecord]() // one per UMI group
+  private var readIdx = 1
+
+  /** True if there are more consensus reads, false otherwise. */
+  def hasNext(): Boolean = {
+    if (this.nextConsensusRecords.nonEmpty) true
+    else this.advance()
+  }
+
+  /** Returns the next consensus read. */
+  def next(): SAMRecord = {
+    if (!this.hasNext()) throw new IllegalStateException("Calling next() when hasNext() is false.")
+    this.nextConsensusRecords match {
+      case mutable.Queue(x, _*) => this.nextConsensusRecords.dequeue()
+      case mutable.Queue() => unreachable("hasNext() was true but nextConsensusRecords is empty")
+    }
+  }
+
+  /** Consumes records until a consensus read can be created, or no more input records are available. Returns
+    * true if a consensus read was created, false otherwise. */
+  private def advance(): Boolean = {
+    if (this.iter.isEmpty) return false
+
+    // get the records to create the consensus read
+    val buffer = getNextGroupOfRecords(attr=this.iter.head.toAttr)
+
+    // partition the records to which end of a pair it belongs, or if it is a fragment read.
+    val (fragments, firstOfPair, secondOfPair) = subGroupRecords(records = buffer)
+
+    // build a consensus for each sub-group that has a read
+    val fragmentName = nextReadName(fragments)
+    val pairName = nextReadName(firstOfPair ++ secondOfPair) // must have the same read name
+    var success = false
+    success |= this.callConsensus(fragments,    readName=fragmentName, readType=Fragment)
+    success |= this.callConsensus(firstOfPair,  readName=pairName,     readType=FirstOfPair)
+    success |= this.callConsensus(secondOfPair, readName=pairName,     readType=SecondOfPair)
+
+    if (success) true
+    else if (this.iter.nonEmpty) this.advance() // try again
+    else false // failed to build a consensus, and no more records in the input iterator
+  }
+
+  /** Adds a SAM record from the underlying iterator to the buffer if either the buffer is empty or the SAM attribute is
+    * the same for the records in the buffer as the next record in the input iterator.  Returns true if a record was
+    * added, false otherwise.
+    */
+  @annotation.tailrec
+  private def getNextGroupOfRecords(buffer: ListBuffer[SAMRecord] = new ListBuffer[SAMRecord], attr: String): ListBuffer[SAMRecord] = {
+    if (this.iter.isEmpty || this.iter.head.toAttr != attr) return buffer
+    val rec = this.iter.next()
+    progress.map(_.record(rec))
+    buffer.append(rec)
+    getNextGroupOfRecords(buffer=buffer, attr=attr)
+  }
+
+  /** Adds a method to the SAMRecord class to retrieve the attribute for the UMI identifier. */
+  implicit private class SAMRecordToString(r: SAMRecord) {
```

Comment:
Is this really necessary?  I see three usages that could easily be replaces with the inlined code without causing all the extra overhead.